ADAL/src/workplacejoin/mac/ADWorkPlaceJoinUtil.m

```diff
@@ -22,14 +22,292 @@
+                                                            privateKey:privateKey];
+    SAFE_ARC_AUTORELEASE(info);
+    
+    // Fall through to clean up resources.
+    
+_error:
+    
+    SAFE_ARC_RELEASE(certificateSubject);
+    SAFE_ARC_RELEASE(certificateData);
+    
+    if (identity)
+        CFRelease(identity);
+    if (certificate)
+        CFRelease(certificate);
+    if (privateKey)
+        CFRelease(privateKey);
+    
+    return info;
+}
+
+
++ (void) copyCertificate:(NSUUID*)correlationId
+                identity:(SecIdentityRef __nullable * __nonnull)identity
+             certificate:(SecCertificateRef __nullable * __nonnull)clientCertificate
+       certificateIssuer:(NSString* __nullable * __nonnull)clientCertificateIssuer
+                   error:(ADAuthenticationError * __nullable __autoreleasing * __nullable)error
+{
+    OSStatus status = noErr;
+    ADAuthenticationError* adError = nil;
+    NSData* issuer = nil;
+    NSMutableDictionary *identityQuery = nil;
+    CFDictionaryRef result = NULL;
+    
+    *identity = nil;
+    *clientCertificate = nil;
+    if (error)
+        *error = nil;
+
+    *clientCertificate = [self copyWPJCertificateRef:correlationId error:&adError];
+    
+    if (adError)
+    {
+        if (error)
+            *error = adError;
+        
+        AD_LOG_ERROR_F(@"Failed to retrieve WPJ client certificate from keychain - ", adError.code, nil, @"%@ correlation id.", correlationId);
+        goto _error;
+    }
+    
+    // If there's no certificate in the keychain, adError won't be set since this isn't an error condition.
+    if (!*clientCertificate)
+        return;
+    
+    // In OS X the shared access group cannot be set, so the search needs to be more
+    // specific. The code below searches the identity by passing the WPJ cert as reference.
+    identityQuery = [[NSMutableDictionary alloc] init];
+    SAFE_ARC_AUTORELEASE(identityQuery);
+    
+    [identityQuery setObject:(__bridge id)kSecClassIdentity forKey:(__bridge id)kSecClass];
+    [identityQuery setObject:(__bridge id)kCFBooleanTrue forKey:(__bridge id<NSCopying>)(kSecReturnRef)];
+    [identityQuery setObject:(__bridge id)kCFBooleanTrue forKey:(__bridge id<NSCopying>)(kSecReturnAttributes)];
+    [identityQuery setObject:(__bridge id)kSecAttrKeyClassPrivate forKey:(__bridge id)kSecAttrKeyClass];
+    [identityQuery setObject:(__bridge id)*clientCertificate forKey:(__bridge id)kSecValueRef];
+    
+    status = SecItemCopyMatching((__bridge CFDictionaryRef)identityQuery, (CFTypeRef*)&result);
+    CHECK_KEYCHAIN_STATUS(@"Failed to retrieve WPJ identity from keychain.");
+    
+    issuer = [(__bridge NSDictionary*)result objectForKey:(__bridge id)kSecAttrIssuer];
+    if (issuer)
+    {
+        *clientCertificateIssuer = [[NSString alloc] initWithData:issuer encoding:NSISOLatin1StringEncoding];
+        SAFE_ARC_AUTORELEASE(*clientCertificateIssuer);
```

Comment:
copy in this method signature means that the calling method should own all of the passed back parameters. The caller of this method should not have to guess which ones they own and which they don't.